tests/memex/storage_test.py

```diff
@@ -407,10 +434,10 @@ def fetch_annotation(patch):
 
 
 @pytest.fixture
-def models(patch):
-    models = patch('memex.storage.models', autospec=False)
-    models.Annotation.return_value.is_reply = False
-    return models
+def patched_annotation(patch):
+    klass = patch('memex.storage.models.Annotation')
+    klass.return_value.is_reply = False
+    return klass
```

Comment:
This seems to work for me, on master:

```diff
diff --git a/tests/memex/storage_test.py b/tests/memex/storage_test.py
index ece6cb5..32863fb 100644
--- a/tests/memex/storage_test.py
+++ b/tests/memex/storage_test.py
@@ -94,8 +94,7 @@ class TestExpandURI(object):
         ]
 
 
-@pytest.mark.usefixtures('models',
-                         'update_document_metadata')
+@pytest.mark.usefixtures('models')
 class TestCreateAnnotation(object):
 
     def test_it_fetches_parent_annotation_for_replies(self,
@@ -181,8 +180,8 @@ class TestCreateAnnotation(object):
 
     def test_it_updates_the_document_metadata_from_the_annotation(self,
                                                                   models,
-                                                                  pyramid_request,
-                                                                  update_document_metadata):
+                                                                  pyramid_request):
+        models.update_document_metadata.assert_called_once_with('foo')
         annotation_data = self.annotation_data()
         annotation_data['document']['document_meta_dicts'] = (
             mock.sentinel.document_meta_dicts)
@@ -191,7 +190,7 @@ class TestCreateAnnotation(object):
 
         storage.create_annotation(pyramid_request, annotation_data)
 
-        update_document_metadata.assert_called_once_with(
+        models.update_document_metadata.assert_called_once_with(
             pyramid_request.db,
             models.Annotation.return_value,
             mock.sentinel.document_meta_dicts,
```

The added assertion fails as I'd expect it to. (And if I remove the added assertion then the test passes as I'd expect it to.)

Or in other words: if you've already patched `model` then you don't need to patch `update_document_metadata`, just use the `model.update_document_metadata` of your existing `model` mock.

The reason I'm keen on this is that `patch('memex.storage.models')` patches a top-level name in the module under test (`memex.storage`). _Only_ `memex.storage` will see this mock object in place of `models`, any other code that happens to be run during the test and that also uses `memex.storage.models` will see the real `models` module.

Whereas `patch('memex.storage.models.update_document_metadata')` will replace the `update_document_metadata()` function in the `memex.storage.models` module with a mock, and _any_ code run during the test that is accessing `memex.storage.models.update_document_metadata()` will see a mock instead of the real function, _not_ just the code under test.

I don't think that creates a problem in this case, but it has created difficult to debug issues elsewhere, for example [breaking the test factories](https://github.com/hypothesis/h/pull/3850/commits/bb7fbb03adfede15db249fe09912cbcc8810a9af). So I think it's a good idea to _always patch names on the module under test's own namespace, and never patch names within objects imported into the module under test_.